### PR TITLE
samples/subsys: add sample for the settings

### PR DESCRIPTION
**Pull request by nvlsianpu**
<b>Added sample for the settings subsystem.</b>

The sample shows how to:
- initialize and register handler
- implement handles
- save and load data using registered handlers
- load subtree
- save or delete a certain value
- load subtree values or a value directly
- example on how to write data to the
setting destination and how to read data
from the setting destination using runtime API

Sample is able to be run on qemu_x86. native_posix and nrf52 targets using booth NVS and FS back-ends.

<b>Additionally:</b>
Corrected the setiings subsystem doc:
- Added return value description to each of settings handler
description.
- Include run-time API into doxygen build.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

### CODE REVIEW

**Comment by nvlsianpu** [comment]
cc @carlescufi @rakons

File: doc/reference/runtime_conf/index.rst [hunk 1]
```diff
@@ -15,6 +15,9 @@ storage medium, and even change it later as needs change.
 The runtime configuration system is used by various Zephyr
 components and can be used simultaneously by user applications.
 
+For pure example of the settings module refer to
```
Comment (carlescufi) [inline doc/reference/runtime_conf/index.rst]: For an example of the settings module, refer to...

File: doc/reference/runtime_conf/index.rst [hunk 2]
```diff
@@ -15,6 +15,9 @@ storage medium, and even change it later as needs change.
...
+:ref:`the sample <settings_subsy_sample>`.
```
Comment (carlescufi) [inline doc/reference/runtime_conf/index.rst]: `settings_subsys_sample`

File: samples/subsys/settings/README.rst [hunk 3]
```diff
@@ -0,0 +1,100 @@
+.. _settings_subsy_sample:
```
Comment (carlescufi) [inline samples/subsys/settings/README.rst]: `settings_subsys_sample`

**Review by carlescufi — CHANGES_REQUESTED** [review]
Just a couple of naming changes

**Comment by nvlsianpu** [comment]
@carlescufi updated

**Review by carlescufi — APPROVED** [review]

**Comment by nvlsianpu** [comment]
cc @lemrey @kapi-no

**Comment by nvlsianpu** [comment]
@de-nordic maybe you can find some time to review?

File: samples/subsys/settings/prj.conf [hunk 4]
```diff
@@ -0,0 +1,8 @@
+CONFIG_STDOUT_CONSOLE=y
+CONFIG_FLASH=y
+CONFIG_FLASH_PAGE_LAYOUT=y
+CONFIG_FLASH_MAP=y
+
+CONFIG_SETTINGS=y
+CONFIG_SETTINGS_RUNTIME=y
+CONFIG_SETTINGS_USE_BASE64=n
```
Comment (de-nordic) [inline samples/subsys/settings/prj.conf]: I do not think we need BASE64 anymore.

File: samples/subsys/settings/native_posix.overlay [hunk 5]
```diff
@@ -0,0 +1,25 @@
+/*
```
Comment (de-nordic) [inline samples/subsys/settings/native_posix.overlay]: I do not think this file is needed anymore, I have been able to run this sample without it. Can you check it?

File: samples/subsys/settings/native_posix_64.overlay [hunk 6]
```diff
@@ -0,0 +1,25 @@
+/*
```
Comment (de-nordic) [inline samples/subsys/settings/native_posix_64.overlay]: ... also can you check if this file is still needed?

**Review by kapi-no — APPROVED** [review]

File: samples/subsys/settings/sample.yaml [hunk 7]
```diff
@@ -0,0 +1,33 @@
+sample:
+  name: Settings Sample
+
+tests:
+  sample.subsys.settings:
+    tags: settings
+    depends_on: nvs
+    timeout: 10
+    platform_whitelist: qemu_x86
```
Comment (de-nordic) [inline samples/subsys/settings/sample.yaml]: can we add native_posix and native_posix_64 ?

Comment (nvlsianpu) [reply samples/subsys/settings/sample.yaml]: added

File: samples/subsys/settings/prj.conf (hunk 4 above)
Comment (nvlsianpu) [reply samples/subsys/settings/prj.conf]: yes, removed

**Review by de-nordic — APPROVED** [review]

File: samples/subsys/settings/sample.yaml [hunk 8]
```diff
@@ -0,0 +1,33 @@
+sample:
+  name: Settings Sample
+
+tests:
+  sample.subsys.settings:
+    tags: settings
+    depends_on: nvs
```
Comment (nashif) [inline samples/subsys/settings/sample.yaml]: Remove that please, this limits building/testing to qemu_x86 which is the only platform that has nvs set. You either limit by dependency or by whitelisting, not both. Right now the test fails on native_posix.

**Review by nashif — CHANGES_REQUESTED** [review]

File: samples/subsys/settings/sample.yaml (hunk 8 above)
Comment (nvlsianpu) [reply samples/subsys/settings/sample.yaml]: Maybe better to remove this dependency? Sample is able to run using NVS and FS as bak-ends.

Comment (nvlsianpu) [reply samples/subsys/settings/sample.yaml]: @nashif I removed this line `depends_on: nvs`

Comment (nvlsianpu) [reply samples/subsys/settings/sample.yaml]: I see - the issue is caused by different key read-out order in for FS back-end - its expected behavior, need to fix the harness.

Comment (nvlsianpu) [reply samples/subsys/settings/sample.yaml]: fixed

**Comment by nvlsianpu** [comment]
@nashif I fixed tests, now even native_posix can pass.

**Comment by nvlsianpu** [comment]
@nashif @carlescufi Any chance to have this in 2.2 release?

**Comment by carlescufi** [comment]
@nashif please revisit and merge if happy

**Review by nashif — APPROVED** [review]

**Pull request merged**